### PR TITLE
feat: add tx status in mempool

### LINKED ISSUE #1281: Provide the current status of a transaction for clients
### Summary

@ninabarbakadze recently did some phenomenal work on [feat: add TxStatus rpc endpoint](https://github.com/celestiaorg/celestia-core/commit/e8724248c7d6a438ec08046e82525c83a2d64941). This preliminary work indexed committed transactions so users could see that they were committed in a lightweight manner and not need to rely on `BroadcastTxCommit` or the `Tx` endpoint + the `kv_indexer`.

As part of providing greater tracing ability for users transactions is to modify the `TxStatus` endpoint such that it gives the current status of the transaction. They can be as follows:

- `Unknown`(default case)
- `Pending` (in the mempool)
- `Committed` (in the secondary index in the block store)
- `Rejected` (as decided by the mempools rejected cache
- `Evicted` (has been kicked out of the mempool because of a TTL or not a high enough fee)

### PR DESCRIPTION
**Pull request by ninabarbakadze**
## Description

Fixes #1281 

Opens https://github.com/celestiaorg/celestia-core/issues/1381

### CODE REVIEW

File: mempool/cache.go
Comment (rootulp) [inline mempool/cache.go:20]: [blocking]

```suggestion
	// Push adds the given tx key to the cache and returns true if it was
	// newly added. Otherwise, it returns false.
```

File: mempool/cache.go
Comment (rootulp) [inline mempool/cache.go:23]:
```suggestion
	// Remove removes the given tx key from the cache.
```

File: mempool/cache.go
Comment (rootulp) [inline mempool/cache.go:27]:
```suggestion
	// Has reports whether tx key is present in the cache. Checking for presence is
	// not treated as an access of the value.
```

File: mempool/cache.go
Comment (rootulp) [inline mempool/cache.go:91]: [question] `Remove` is now redundant with `RemoveTxByKey`. Should we remove one of them?

File: mempool/cat/pool.go
Comment (rootulp) [inline mempool/cat/pool.go:188]:
```suggestion
// the specified key was recently evicted and is currently within the cache.
```

File: mempool/mempool.go
Comment (rootulp) [inline mempool/mempool.go:96]: [perhaps clearer]

```suggestion
	// GetTxByKey gets a tx by its key from the mempool cache. Returns the tx and a bool indicating its presence in the tx cache.
	// Used in the RPC endpoint: TxStatus.
```

File: mempool/mock/mempool.go
Comment (rootulp) [inline mempool/mock/mempool.go:10]:
```suggestion
// Mempool is a mock implementation of a Mempool, useful for testing.
```

File: mempool/mock/mempool.go
Comment (rootulp) [inline mempool/mock/mempool.go:35]: [question] why does removing always add the tx to the evictedTxs? Can't a tx be removed from the mempool by accepting it in a block and not evicting it from the mempool?

File: mempool/v0/clist_mempool.go
Comment (rootulp) [inline mempool/v0/clist_mempool.go:186]:
```suggestion
// GetTxByKey is not supported by the v0 mempool but it is required to satisfy the mempool interface.
```

File: mempool/v0/clist_mempool.go
Comment (rootulp) [inline mempool/v0/clist_mempool.go:191]:
```suggestion
// GetTxEvicted is not supported by the v0 mempool but it is required to satisfy the mempool interface.
```

File: mempool/v1/mempool.go
Comment (rootulp) [inline mempool/v1/mempool.go:277]:
```suggestion
// the specified key was recently evicted and is currently within the cache.
```

File: rpc/core/tx_status.go
Comment (rootulp) [inline rpc/core/tx_status.go:13]:
```suggestion
// or whether the transaction is pending, evicted from the mempool, or otherwise unknown.
```

File: rpc/core/tx_status.go
Comment (rootulp) [inline rpc/core/tx_status.go:42]: [question] should we extract consts for `PENDING`, `EVICTED`, `COMMITTED`, and `UNKNOWN`?

**Review by rootulp — APPROVED** [review]
LGTM. Suggestions were mostly optional comment improvements

File: rpc/core/tx_status.go
Comment (cmwaters) [inline rpc/core/tx_status.go:1]: There is already a `tx.go` file. I think it makes sense to have this method there (next to the the transaction and txSearch methods)

File: mempool/v1/mempool.go
Comment (cmwaters) [inline mempool/v1/mempool.go:220]:
```suggestion
		if !txmp.cache.Push(txKey) {
```
The key was already computed above

File: mempool/v1/mempool.go
Comment (cmwaters) [inline mempool/v1/mempool.go:238]: Same as above. There is already a `txKey` you can use

File: mempool/cat/pool.go
Comment (cmwaters) [inline mempool/cat/pool.go:179]: This is a breaking change. It may be better to still have `Get` and make it wrap `GetTxByKey` if you want to change the name of the method

File: mempool/mock/mempool.go
Comment (cmwaters) [inline mempool/mock/mempool.go:14]: It might be better to use something like mockery and have generated mocks then to reuse this empty implementation. It also means that you've had to touch a lot of different files

File: rpc/core/tx_status.go
Comment (cmwaters) [reply rpc/core/tx_status.go:42]: Yes we should, or even use an `enum`

File: rpc/core/tx_status.go
Comment (ninabarbakadze) [reply rpc/core/tx_status.go:1]: I'm putting it there but i couldn't find where those methods were tested so i kept the test file.

File: mempool/v1/mempool.go
Comment (ninabarbakadze) [reply mempool/v1/mempool.go:238]: txKey is computed in the anonymous function above

File: mempool/cache.go
Comment (ninabarbakadze) [reply mempool/cache.go:91]: would that no be api breaking?

File: mempool/mock/mempool.go
Comment (ninabarbakadze) [reply mempool/mock/mempool.go]: i only did this in the mock to test txstatus which now has been reverted

File: mempool/cache.go
Comment (rootulp) [reply mempool/cache.go:91]: good point, let's keep both if they were already part of the public API already

File: mempool/v0/clist_mempool.go
Comment (rootulp) [inline mempool/v0/clist_mempool.go:191]: [question] why is it difficult to add support for `GetTxEvicted` to the v0 mempool?

File: mempool/v0/clist_mempool.go
Comment (rootulp) [inline mempool/v0/clist_mempool.go:186]: [question] why is it difficult to add support for GetTxByKey to the v0 mempool? Isn't this just a wrapper around the already existing `GetTx`?

File: rpc/client/rpc_test.go
Comment (rootulp) [inline rpc/client/rpc_test.go:539]: [optional] should this include an assertion or test case for evicted transactions?

File: rpc/core/tx.go
Comment (rootulp) [inline rpc/core/tx.go:243]: note: another sign that the method should be renamed from `GetTxEvicted` to `IsTxEvicted`

**Review by rootulp — COMMENTED** [review]
Great work! I have one blocking question about breaking the interface

File: mempool/cache.go
Comment (cmwaters) [reply mempool/cache.go:28]: It's not necessary in the sense that it can be achieved without breaking it however we're needlessly computing the hash multiple times and this isn't an interface that we expect users to use. I don't think users can even use it in the sense that the mempool constructor doesn't accept a TxCache interface so this interface would be better of being private (and used just for testing)

File: rpc/core/tx.go
Comment (cmwaters) [reply rpc/core/tx.go:243]: agree, "get" implies the type is returned, "is" implies a boolean is returned

File: rpc/core/tx.go
Comment (cmwaters) [inline rpc/core/tx.go:227]: As a follow up issue: we will probbaly need to allow for the user to request a proof of the tx's inclusion if it has been committed

File: mempool/v1/mempool.go
Comment (cmwaters) [inline mempool/v1/mempool.go:562]:
```suggestion
			txmp.cache.Remove(wtx.key)
```

File: mempool/v1/mempool.go
Comment (cmwaters) [inline mempool/v1/mempool.go:573]:
```suggestion
			txmp.evictedTxs.Push(wtx.key)
```

File: mempool/v1/mempool.go
Comment (cmwaters) [reply mempool/v1/mempool.go:562]: The same can be applied to all the other areas that `wtx` is used

File: rpc/core/tx_status_test.go
Comment (cmwaters) [inline rpc/core/tx_status_test.go:24]: :+1: Thanks

File: mempool/cache_bench_test.go
Comment (cmwaters) [inline mempool/cache_bench_test.go:23]: nit: This is a slightly incorrect way of performing the benchmark. The `Key()` method should be done outside of the benchmark the same way that the constructor is made. In fact you just need to make random 32 byte slices (i.e. they don't have to be actual hashes so long as they are unique)

File: mempool/cat/pool.go
Comment (ninabarbakadze) [reply mempool/cat/pool.go:178]: thank you!

File: rpc/client/rpc_test.go
Comment (ninabarbakadze) [reply rpc/client/rpc_test.go:539]: yes should be good!

File: rpc/core/tx.go
Comment (ninabarbakadze) [reply rpc/core/tx.go:227]: https://github.com/celestiaorg/celestia-core/issues/1389

File: mempool/v0/clist_mempool.go
Comment (rootulp) [reply mempool/v0/clist_mempool.go:191]: sorry for repeating myself but why? I don't have enough info to understand why we're deprecating it instead of adding this feature and maintaining it. It hasn't been a maintenance burden before this feature so I'm wondering why adding support for `GetTxEvicted` is difficult for the v0 mempool.

File: mempool/cache.go
Comment (rootulp) [reply mempool/cache.go:28]: Fair enough. Given the interface is public this PR still seems breaking so we should consider adding `!` to the PR title. IMO we could've split this into two separate PRs:
1. Add the feature and don't break the interface.
2. A separate PR for breaking the interface to reduce hashing if that's really a problem.

Since this PR targets celestia-core and we don't know how to release breaking changes of this repo, I'm concerned this PR won't get released from `main` anytime soon. Also if it's technically API breaking, I don't think we're supposed to backport v0.34.x-celestia?

If we split this into a non-breaking change and a breaking change, we can easily backport the non-breaking change.

File: rpc/client/rpc_test.go
Comment (ninabarbakadze) [reply rpc/client/rpc_test.go:539]: so, the mempool there uses default mempool config where the cache size is 5000 txs. I couldn't overflow that in this unit test and also didn't see a way of changing the mempool. 

any suggestions welcome

File: mempool/cache.go
Comment (cmwaters) [reply mempool/cache.go:28]: It's a bit silly that this has to be breaking. As I alluded to, you can't actually use the interface for anything (you don't get to choose what cache the mempool uses) and it should've been private (if we had wanted it for testing purposes). Since this is only performative and we want this feature in v2 not v3, @ninabarbakadze can you revert these changes to make it non-breaking. If we want we can revisit this later on

**Review by evan-forbes — APPROVED** [review]
LGTM

File: mempool/cache.go
Comment (ninabarbakadze) [reply mempool/cache.go:28]: reverting the changes

File: mempool/v0/clist_mempool.go
Comment (ninabarbakadze) [reply mempool/v0/clist_mempool.go]: Sorry for the confusion. The idea of deprecating v0 mempool came from having to make changes to v0 every time we decided to alter the mempool interface. I don't think that it's difficult to maintain rather why maintain it if there's a possibility to remove something we're not currently supporting.

Comment (rootulp) [reply mempool/v0/clist_mempool.go]: > The idea of deprecating v0 mempool came from having to make changes to v0 every time we decided to alter the mempool interface.

Seems like we can do this PR without altering the mempool interface as suggested in a different comment 

> if there's a possibility to remove something we're not currently supporting.

We currently support the v0 mempool. In fact, it's the default on many other Cosmos chains because the v1 (prioritized) mempool was deprecated.

File: mempool/cat/store.go
Comment (rootulp) [inline mempool/cat/store.go:145]: [no change needed] it seems like `purge` could be renamed to `evict` here. IMO it's easier to understand what's happening if we use consistent names for consistent things.

In other words, it isn't immediately obvious to me that purging a transaction should add it to the evicted tx cache. However, evicting a transaction seems like it should add the tx to the evicted tx cache.

File: mempool/mempool.go
Comment (rootulp) [inline mempool/mempool.go:94]: [question] should this clarify that it's referring to the mempool tx cache? Or get rid of cache entirely? "cache" seems ambiguous because there is now a tx cache and an evicted tx cache.

```suggestion
	// GetTxByKey gets a tx by its key from the mempool tx cache. Returns the tx and a bool indicating its presence in the tx cache.
```

```suggestion
	// GetTxByKey gets a tx by its key from the mempool. Returns the tx and a bool indicating its presence in the tx cache.
```

File: mempool/mempool.go
Comment (rootulp) [inline mempool/mempool.go:98]: nit

```suggestion
	// IsTxEvicted returns true if the tx was evicted from the mempool and exists in the
```

Comment (rootulp) [reply mempool/mempool.go:98]: [question] what happens if the tx gets evicted from the mempool and then the user re-submits the exact same tx so that it exists in the tx cache? Is it possible for one tx to exist in both the tx cache and the evicted tx cache at the same time?

File: mempool/v0/clist_mempool.go
Comment (rootulp) [reply mempool/v0/clist_mempool.go:186]: [blocking] I don't think we have rough consensus on deprecating the v0 mempool. See https://github.com/celestiaorg/celestia-core/issues/1382.

IMO if it's easy to add, then we should add support for `GetTxByKey` until we make a decision on deprecating the v0 mempool. If it's difficult to add support for `GetTxByKey` to the v0 mempool, that should be information that we use when making the decision to deprecate or not. If that's the case, making the decision should block merging this PR.

File: mempool/v0/clist_mempool.go
Comment (rootulp) [inline mempool/v0/clist_mempool.go:194]: [blocking] I don't think we have rough consensus on deprecating the v0 mempool. See https://github.com/celestiaorg/celestia-core/issues/1382.

IMO if it's easy to add, then we should add support for `IsTxEvicted` until we make a decision on deprecating the v0 mempool. If it's difficult to add support for `IsTxEvicted` to the v0 mempool, that should be information that we use when making the decision to deprecate or not. If that's the case, making the decision should block merging this PR.

File: mempool/v1/mempool.go
Comment (rootulp) [inline mempool/v1/mempool.go:277]:
```suggestion
// IsTxEvicted returns a bool indicating whether the transaction with
// the specified key was recently evicted and is currently within the evicted cache.
```

File: mempool/v0/clist_mempool.go
Comment (ninabarbakadze) [reply mempool/v0/clist_mempool.go:194]: if i'm understanding it correctly v0 is currently not in use. I'm not sure why we'd want to support TxStatus in v0 if it's not enabled and won't be used. We could merge this pr as long as we satisfy the interface. 

I have no problem implementing those methods in v0 I just don't really understand why we should

File: mempool/mempool.go
Comment (ninabarbakadze) [reply mempool/mempool.go]: yes, it's possible that a previously evicted tx exists in both caches. evicted tx only gets rid of its records once it reaches its max size(v0) or once it's reset alongside all other caches(v1). I think how we can avoid confusion between re-added txs is if we check for eviction last.

Comment (cmwaters) [reply mempool/mempool.go]: I think a more correct phrasing of the method is `WasRecentlyEvicted`

File: mempool/cat/store.go
Comment (cmwaters) [inline mempool/cat/store.go:145]: I don't like the cache been passed into the store. It feels leaky. Either the caches are built into the store or this method returns an array of evicted tx keys that the mempool can then plug into the cache. I think I would prefer the latter

File: mempool/v0/clist_mempool.go
Comment (cmwaters) [reply mempool/v0/clist_mempool.go:194]: The point I think rootul is making is even although it's not the default, it is feasible that someone use the mempool v0, thus we do support it and until we stop supporting it, features like `TxStatus` will need to be implemented across all implementations

**Review by cmwaters — COMMENTED** [review]
This is almost done, just a few remaining remarks

**Review by ninabarbakadze — COMMENTED** [review]
Added txstatus in v0 @cmwaters @rootulp

File: mempool/cat/store.go
Comment (ninabarbakadze) [reply mempool/cat/store.go]: I returns purged txs from the store but i'm afraid that's also api breaking?

File: mempool/cat/store.go
Comment (rootulp) [inline mempool/cat/store.go:144]:
```suggestion
// and time. Returns the purged txs and amount of transactions that were purged.
```

still seems like purge could be renamed to evicted

File: mempool/v0/clist_mempool.go
Comment (rootulp) [inline mempool/v0/clist_mempool.go:196]: [question] why does this implementation not support transaction eviction?

**Review by rootulp — APPROVED** [review]

File: mempool/v0/clist_mempool.go
Comment (ninabarbakadze) [reply mempool/v0/clist_mempool.go:196]: evicted txs are valid transactions that were evicted because of size/ttl reasons for example when mempool is full. afaik Clist mempool on the other hand directly rejects transactions if the cache is full rather than evicting them.

Comment (cmwaters) [reply mempool/v0/clist_mempool.go:196]: There is also no TTL on the v0 mempool

**Review by cmwaters — APPROVED** [review]
LGTM - I have one minor comment which you can choose to do as a follow up PR if you wish

File: rpc/core/tx.go
Comment (ninabarbakadze) [reply rpc/core/tx.go:234]: no it doesn't I think there should be an issue for this otherwise i'll create one

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**
